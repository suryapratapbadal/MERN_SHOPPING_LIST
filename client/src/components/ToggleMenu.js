import React from 'react';
import IconButton from '@material-ui/core/IconButton';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import MoreVertIcon from '@material-ui/icons/MoreVert';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { updateItem } from '../actions/index';

const ITEM_HEIGHT = 25;

export class ToggleMenu extends React.Component {
  state = {
    anchorEl: null,
    shopped: this.props.completed 
  };

  handleClick = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  componentDidUpdate (prevProps) {
      if(prevProps.completed !== this.props.completed)
      this.setState({shopped : this.props.completed})
  }

  handleClose = option => {
      
      if(option === 'Delete') {
          this.props.toggleDeleteConfirmationDialogue();
          this.setState({ anchorEl: null });
      } else if(option === 'Done' || option === 'Undone') {
          this.props.updateItem(this.props.id,{"completed": !this.state.shopped})
          this.setState({ anchorEl: null})
      } else {
        this.setState({ anchorEl: null})
      }
    
  };

  render() {
    const { anchorEl } = this.state;
    const open = Boolean(anchorEl);
    const options = [
        'Delete',
        this.state.shopped ? 'Undone' : 'Done'
      ];

    return (
      <div>
        <IconButton
          aria-label="More"
          aria-owns={open ? 'long-menu' : null}
          aria-haspopup="true"
          onClick={this.handleClick}
        >
          <MoreVertIcon />
        </IconButton>
        <Menu
          id="long-menu"
          anchorEl={anchorEl}
          open={open}
          onClose={this.handleClose}
          PaperProps={{
            style: {
              maxHeight: ITEM_HEIGHT * 4.5,
              width: 100,
            },
          }}
        >
          {options.map(option => (
            <MenuItem key={option} selected={option === 'Delete'} onClick={this.handleClose.bind(this,option)}>
              {option}
            </MenuItem>
          ))}
        </Menu>
      </div>
    );
  }
}

export default connect(state => {
    const loading = state.itemReducer.loading || false;
    return {
        loading
    }
}, dispatch => {
    return bindActionCreators({ updateItem: updateItem }, dispatch)
}
)(ToggleMenu);