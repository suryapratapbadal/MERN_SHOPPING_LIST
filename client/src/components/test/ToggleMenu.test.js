import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// Component to be tested
import { ToggleMenu } from '../ToggleMenu';


const mockupdatefn = jest.fn();
const toggleDeleteConfirmationDialoguefn = jest.fn();
const id = '1';


// Component Renders
describe('<ToggleMenu />', () => {

    const wrapper = shallow(<ToggleMenu updateItem={mockupdatefn} loading={false} toggleDeleteConfirmationDialogue={toggleDeleteConfirmationDialoguefn} id={id} completed={false}/>);
    const event = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn(),
        currentTarget: 'dummy'
    }
    describe('render()', () => {

        test('renders the component', () => {

            expect(toJson(wrapper)).toMatchSnapshot();

        });
    });


    describe('onClick()',()=>{
        test('handle click on button',()=>{
            wrapper.find('WithStyles(IconButton)').props().onClick(event);
            expect(wrapper.state('anchorEl')).toEqual('dummy');
        });
    });

    describe('onClose()',()=>{
        test('handle click on button',()=>{
            wrapper.find('WithStyles(Menu)').props().onClose();
            expect(wrapper.state('anchorEl')).toEqual(null);
        });
    });

    describe('onClick()',()=>{
        test('handle click on button',()=>{
            wrapper.find('WithStyles(MenuItem)').at(0).props().onClick('Done');
            expect(wrapper.state('anchorEl')).toEqual(null);
        });
    });
});