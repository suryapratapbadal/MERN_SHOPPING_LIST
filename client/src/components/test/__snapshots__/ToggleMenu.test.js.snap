// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleMenu /> render() renders the component 1`] = `
<div>
  <WithStyles(IconButton)
    aria-haspopup="true"
    aria-label="More"
    aria-owns={null}
    onClick={[Function]}
  >
    <pure(MoreVertIcon) />
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    PaperProps={
      Object {
        "style": Object {
          "maxHeight": 112.5,
          "width": 100,
        },
      }
    }
    anchorEl={null}
    id="long-menu"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(MenuItem)
      key="Delete"
      onClick={[Function]}
      selected={true}
    >
      Delete
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="Done"
      onClick={[Function]}
      selected={false}
    >
      Done
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
</div>
`;
